reactive/kotlinx-coroutines-rx2/src/RxConvert.kt

```diff
@@ -77,6 +78,31 @@ public fun <T : Any> ReceiveChannel<T>.asObservable(context: CoroutineContext):
         send(t)
 }
 
+/**
+ * Transforms given cold [ObservableSource] into cold [Flow].
+ *
+ * The resulting flow is _cold_, which means that [ObservableSource.subscribe] is called every time a terminal operator
+ * is applied to the resulting flow.
+ *
+ * A channel with the [default][Channel.BUFFERED] buffer size is used. Use the [buffer] operator on the
+ * resulting flow to specify a user-defined value and to control what happens when data is produced faster
+ * than consumed, i.e. to control the back-pressure behavior. Check [callbackFlow] for more details.
+ */
+@ExperimentalCoroutinesApi
+public fun <T: Any> ObservableSource<T>.asFlow(): Flow<T> = callbackFlow {
+
+    var disposable: Disposable? = null
+
+    val observer = object : Observer<T> {
+        override fun onComplete() { close() }
+        override fun onSubscribe(d: Disposable) { disposable = d }
+        override fun onNext(t: T) { sendBlocking(t) }
+        override fun onError(e: Throwable) { close(e) }
+    }
+    subscribe(observer)
+    awaitClose { disposable?.dispose() }
```

Comment:
There is a potential data race here. I cannot find a specific example that would demonstrate it, but, at the same time, I cannot find any place in `Observable` documentation where it guarantees that `onSubscribe` would be called synchronously, so potentially it can be called at some later time when `awaitClose { disposable?.dispose() }` had already completed. Moreover, operators in Rx are written with the assumption that this might happen. Take a look at the logic of `DisposableHelper.setOnce`: https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/disposables/DisposableHelper.java#L89-L110
A similar logic is needed here.

Reply:
Done, here:
https://github.com/Kotlin/kotlinx.coroutines/pull/1826